lib/spack/spack/package_base.py

```diff
@@ -432,6 +432,43 @@ def remove_files_from_view(self, view, merge_map):
 
 Pb = TypeVar("Pb", bound="PackageBase")
 
+WhenDict = Dict[spack.spec.Spec, Dict[str, Any]]
+NameValuesDict = Dict[str, List[Any]]
+NameWhenDict = Dict[str, Dict[spack.spec.Spec, List[Any]]]
+
+
+def _by_name(
```

Comment:
If I get this correctly, the function converts to the old `dependencies` format. Where is this needed? Can we avoid it?

Reply:
there are some loops that are easier to do by name, and some that are easier to do by `when`.  It'll be more obvious when we do variants but I think it's needed (and useful) to have both.